packages/flutter/lib/src/widgets/editable.dart

```diff
@@ -43,11 +39,29 @@ class TextRange {
 
   /// Whether this range is empty (but still potentially placed inside the text).
   bool get isCollapsed => start == end;
+
+  /// The text before this range.
+  String textBefore(String text) {
+    return text.substring(0, start);
```

Comment:
I don't know if this is intended but depending on `TextRange` and `text` this can lead to `RangeError` (for instance with `TextRange.empty`, with `TextRange(start: 3, end: 5)` and  `text="aa"`)